Add GetPythonThreadID and Interrupt methods in PythonEngine

gpetrou opened the pull request:
### What does this implement/fix? Explain your changes.

Fixes https://github.com/pythonnet/pythonnet/issues/766 by adding an Interrupt method in PythonEngine class.

### Checklist

Check all those that are applicable and complete.

-   [x] Make sure to include one or more tests for your change
-   [ ] If an enhancement PR, please create docs and at best an example
-   [x] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [x] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

filmor commented on src/embed_tests/TestInterrupt.cs:
```diff
@@ -0,0 +1,83 @@
+
+        [DllImport("pthread", EntryPoint = "pthread_self", CallingConvention = CallingConvention.Cdecl)]
+        private static extern ulong pthread_selfOSX();
+
```
Python 3.8 has `threading.get_native_id()` for this, which we should probably expose in the library with a fallback to these.

gpetrou replied:
I am not sure how to implement this. I have added some changes, what do you think?

filmor commented:
Copying my comment from #1333:

1. This requires a test (done)
2. Strictly, the thread-id attribute is unsigned only for Python >= 3.7 and we still support 3.6
3. Also, since the original type is long, the "correct" type is (U)IntPtr on everything but Windows (Not quite sure how far we have to go there compatibility-wise, the current implementation could be fine for the usual thread IDs)
4. Maybe we should just expose the API function directly (i.e. RaiseInPythonThread on Exception objects or so) and make Interrupt use that

Additional comments on the current state: The helper API to get the current thread id should be in the library and should use `threading.get_native_id()` on Python >=3.8.

filmor commented:
It would also be good if you'd just use normal commits now without force-pushing a single commit over and over, otherwise it becomes very difficult to follow your changes. We can (and will) squash this in the end.

gpetrou commented:
Please note that I am very unfamiliar with the codebase, so I will need more explanations from you. With regards to:
2. I added a version check. Is this what you mean?
3. I added different calls per OS and version. Is this what you mean?
4. I don't understand what that means. Can you elaborate? Is this still needed?

filmor commented on src/runtime/pythonengine.cs:
```diff
@@ -567,6 +576,60 @@ public static void Exec(string code, IntPtr? globals = null, IntPtr? locals = nu
+                return pthread_selfOSX();
+            }
+
+            return 0;
```
Should raise an exception if we can't find a suitable implementation.

filmor commented on src/runtime/pythonengine.cs:
```diff
@@ -567,6 +576,60 @@ public static void Exec(string code, IntPtr? globals = null, IntPtr? locals = nu
+            if (Runtime.PyVersion >= new Version(3, 8))
+            {
+                dynamic threading = Py.Import("threading");
+                return threading.get_native_id();
```
Should get the `GIL`, I think.

gpetrou replied:
Isn't the user supposed to do that when calling the method, as I did in the test?

gpetrou replied:
This does not seem to be working on Ubuntu. Are you sure that this should work? Looking in Cpython tests I see that get_ident is used.

filmor commented:
This goes in the right direction :)

Regarding 2.: Yes, that's what I meant.
Regarding 3.: This looks better, but the Windows type is `uint` and `int`. .NET's `int` is always 32bit, and on Windows `long` is also 32bit on both x86 and x64.

gpetrou commented:
@filmor can this be merged now?

lostmsu commented on src/runtime/runtime.cs:
```diff
@@ -2143,6 +2143,18 @@ internal static void Py_CLEAR(ref IntPtr ob)
         [DllImport(_PythonDll, CallingConvention = CallingConvention.Cdecl)]
         internal static extern int Py_AddPendingCall(IntPtr func, IntPtr arg);
 
+        [DllImport(_PythonDll, EntryPoint = "PyThreadState_SetAsyncExc", CallingConvention = CallingConvention.Cdecl)]
+        internal static extern int PyThreadState_SetAsyncExc37Windows(uint id, IntPtr exc);
+
+        [DllImport(_PythonDll, EntryPoint = "PyThreadState_SetAsyncExc", CallingConvention = CallingConvention.Cdecl)]
+        internal static extern int PyThreadState_SetAsyncExc36Windows(int id, IntPtr exc);
+
+        [DllImport(_PythonDll, EntryPoint = "PyThreadState_SetAsyncExc", CallingConvention = CallingConvention.Cdecl)]
+        internal static extern int PyThreadState_SetAsyncExc37NonWindows(ulong id, IntPtr exc);
+
+        [DllImport(_PythonDll, EntryPoint = "PyThreadState_SetAsyncExc", CallingConvention = CallingConvention.Cdecl)]
+        internal static extern int PyThreadState_SetAsyncExc36NonWindows(long id, IntPtr exc);
+
```
It is better to name these according to [64 bit data models](https://en.wikipedia.org/wiki/LP64#64-bit_data_models) as opposed to `Windows` and `NonWindows`.

lostmsu commented on src/runtime/runtime.cs:
```diff
@@ -2143,6 +2143,18 @@ internal static void Py_CLEAR(ref IntPtr ob)
         internal static extern int Py_AddPendingCall(IntPtr func, IntPtr arg);
 
+        [DllImport(_PythonDll, EntryPoint = "PyThreadState_SetAsyncExc", CallingConvention = CallingConvention.Cdecl)]
+        internal static extern int PyThreadState_SetAsyncExc37Windows(uint id, IntPtr exc);
```
Why do you need overloads with `uint` and `int`? They are basically equivalent at bit level.

gpetrou replied:
I added them because of "Strictly, the thread-id attribute is unsigned only for Python >= 3.7 and we still support 3.6" comment above. Do you want me to remove them?

lostmsu replied:
@filmor on the binary side there's no difference. The caller, if necessary, can bitcast. Any thoughts?

gpetrou replied:
I have removed the int and long declarations for now.

lostmsu commented on src/runtime/pythonengine.cs:
```diff
@@ -567,6 +576,54 @@ public static void Exec(string code, IntPtr? globals = null, IntPtr? locals = nu
+        /// Gets the native thread ID.
+        /// </summary>
+        /// <returns>The native thread ID.</returns>
+        public static ulong GetNativeThreadID()
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return GetCurrentThreadId();
+            }
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+            {
+                return pthread_selfLinux();
+            }
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            {
+                return pthread_selfOSX();
+            }
+
+            throw new InvalidOperationException("Could not retrieve native thread ID.");
+        }
```
I don't think this belongs to `PythonEngine`. From what I read, for the test you might be able to use `threading.get_ident()` from Python `thread` module.

filmor replied:
`get_ident` is a separate thing, the ODs are not compatible with the "remote raise".

lostmsu replied:
@filmor are you sure? Documentation is very bad on this, but here's what I found:

`get_ident` [appears to be](https://github.com/python/cpython/blob/6104013838e181e3c698cb07316f449a0c31ea96/Modules/_threadmodule.c#L1208) returning `PyThread_get_thread_ident`
`tstate->thread_id` [is also](https://github.com/python/cpython/blob/41010184880151d6ae02a226dbacc796e5c90d11/Python/pystate.c#L625) `PyThread_get_thread_ident()`

`PyThreadState_SetAsyncExc` [finds thread by its `tstate->thread_id`](https://github.com/python/cpython/blob/41010184880151d6ae02a226dbacc796e5c90d11/Python/pystate.c#L1129)

gpetrou replied:
Using get_ident seems to work. What would you like me to do? Modify GetNativeThreadID method or remove it completely?

lostmsu replied:
Remove it would be better.

filmor replied:
@lostmsu You are right. But then we should still expose `threading.get_ident` or `PyThread_get_thread_ident`, no?

lostmsu replied:
Why? The user can do what tests do now - e.g. call `Py.Import("threading").Invoke("get_ident")`.

@gpetrou It might make sense to mention `threading.get_ident()` in the summary of XML doc for the `Interrupt` method.

gpetrou replied:
I am getting mixed signals between the two of you :) Who is the BDFL for Python.NET? I added the get_ident usage in GetNativeThreadID for now.

lostmsu replied:
@gpetrou @filmor  well, I am voting for removal, as it is one less thing to support with potential multiple meanings, and having nothing to do with Python embedding on itself.

lostmsu replied:
@filmor ?

lostmsu commented on src/runtime/pythonengine.cs:
```diff
@@ -567,6 +576,54 @@ public static void Exec(string code, IntPtr? globals = null, IntPtr? locals = nu
+        /// </summary>
+        /// <param name="nativeThreadID">The native thread ID.</param>
+        /// <returns>The number of thread states modified; this is normally one, but will be zero if the thread id isn’t found.</returns>
+        public static int Interrupt(ulong nativeThreadID)
```
In C# it is better to use `long` type for public method parameters, as (at least older) VB.NET and some other .NET languages do not support `ulong`.

filmor commented on src/runtime/pythonengine.cs:
```diff
@@ -582,28 +573,8 @@ public static void Exec(string code, IntPtr? globals = null, IntPtr? locals = nu
         /// <returns>The native thread ID.</returns>
         public static ulong GetNativeThreadID()
```
If we keep this, it should be renamed to `GetPythonThreadId` and doesn't this require the GIL in general?

gpetrou replied:
Can you please explain why we need GIL here and not in the Interrupt method? GIL is used for both in the tests already.

filmor replied:
And renaming would still be nice, sorry for the back and forth, I misread the docs.
